File: python/Q98_validate_binary_search_tree.py
#
# @lc app=leetcode id=98 lang=python3
#
# [98] Validate Binary Search Tree
#


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val: int = val
        self.left: TreeNode | None = left
        self.right: TreeNode | None = right


# @lc code=start
class Solution:
    def check_valid(self, root: TreeNode | None, minimum: int, maximum: int) -> bool:
        if not root:
            return True

        if not (minimum < root.val < maximum):
            return False

        left_valid: bool = self.check_valid(root.left, minimum, root.val)
        right_valid: bool = self.check_valid(root.right, root.val, maximum)

        return left_valid and right_valid

    def isValidBST(self, root: TreeNode | None) -> bool:
        return self.check_valid(root, -(2**31) - 1, 2**31)


# @lc code=end

File: rust/src/leetcode/Q98_validate_binary_search_tree.rs
/*
 * @lc app=leetcode id=98 lang=rust
 *
 * [98] Validate Binary Search Tree
 */

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @lc code=start
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    fn subtree_range(root: Option<Rc<RefCell<TreeNode>>>, min: i64, max: i64) -> bool {
        if let Some(node) = root {
            let value: i64 = node.borrow().val as i64;

            if !(min < value && value < max) {
                return false;
            }

            let left_valid = Self::subtree_range(node.borrow().left.clone(), min, value);
            let right_valid = Self::subtree_range(node.borrow().right.clone(), value, max);

            left_valid && right_valid
        } else {
            true
        }
    }

    pub fn is_valid_bst(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        Self::subtree_range(root, i64::MIN, i64::MAX)
    }
}
// @lc code=end
